proof-systems/src/darlin/marlin/src/constraints/test.rs

```diff
@@ -0,0 +1,228 @@
+            .unwrap();
+
+            let pc_verifier_key_gadget =
+                PCG::VerifierKeyGadget::alloc(cs.ns(|| "alloc pc verifier key"), || {
+                    Ok(pc_vk.clone())
+                })
+                .unwrap();
+
+            let proof_gadget =
+                ProofGadget::<G, PC, PCG>::alloc_input(cs.ns(|| "alloc proof"), || {
+                    Ok(proof.clone())
+                })
+                .unwrap();
+
+            let public_inputs = {
+                let mut public_inputs = Vec::new();
+                for (i, input) in wrong_inputs.iter().enumerate() {
+                    public_inputs.push(
+                        NonNativeFieldGadget::alloc_input(
+                            cs.ns(|| format!("alloc public input {}", i)),
+                            || Ok(input),
+                        )
+                        .unwrap(),
+                    )
+                }
+                public_inputs
+            };
+
+            MarlinVerifierGadget::<G, GG, PC, PCG>::verify(
+                cs.ns(|| "proof verification"),
+                &pc_verifier_key_gadget,
+                &verifier_key_gadget,
+                &public_inputs,
+                &proof_gadget,
+            )
+            .unwrap();
+
+            assert!(!cs.is_satisfied());
+        }
+    }
+
+    mod poseidon_fs {
+        use super::*;
+        use algebra::curves::tweedle::dum::{DumJacobian, TweedledumParameters};
+        use algebra::fields::tweedle::Fr;
+        use algebra::Group;
+        use blake2::Blake2s;
+        use fiat_shamir::poseidon::constraints::DensityOptimizedPoseidonQuinticSBoxFSRngGadget;
+        use fiat_shamir::poseidon::TweedleFrPoseidonFSRng;
+        use poly_commit::ipa_pc::{InnerProductArgGadget, InnerProductArgPC};
+        use poly_commit::{
+            DomainExtendedPolyCommitVerifierGadget, DomainExtendedPolynomialCommitment,
+        };
+        use primitives::TweedleFrPoseidonParameters;
+        use r1cs_crypto::TweedleFrDensityOptimizedPoseidonParameters;
+        use r1cs_std::fields::fp::FpGadget;
+        use r1cs_std::groups::curves::short_weierstrass::AffineGadget;
+
+        type G = DumJacobian;
+        type ConstraintF = <G as Group>::BaseField;
+        type GG = AffineGadget<TweedledumParameters, ConstraintF, FpGadget<Fr>>;
+        type FS = TweedleFrPoseidonFSRng;
+        type FSG = DensityOptimizedPoseidonQuinticSBoxFSRngGadget<
+            ConstraintF,
+            TweedleFrPoseidonParameters,
+            TweedleFrDensityOptimizedPoseidonParameters,
+        >;
+        type PC = DomainExtendedPolynomialCommitment<G, InnerProductArgPC<G, FS>>;
+        type PCG = DomainExtendedPolyCommitVerifierGadget<
+            ConstraintF,
+            G,
+            InnerProductArgPC<G, FS>,
+            InnerProductArgGadget<ConstraintF, FSG, G, GG>,
+        >;
+        type D = Blake2s;
+
+        #[test]
+        fn prove_and_verify_with_square_matrix() {
```

Comment:
Can we add another test, in which we actually create and verify a recursive proof (i.e. the one associated to a circuit that verifies another proof) ?

Reply:
Done. In doing so I found out that there were some gadgets not working if the `ConstraintSystem` is set to `SynthesisMode::Setup`. I should have fixed them in commits [01b6444](https://github.com/HorizenOfficial/ginger-lib/pull/174/commits/01b6444d6da31c8e7d1338198391f37528eda764) and [685e6b3](https://github.com/HorizenOfficial/ginger-lib/pull/174/commits/685e6b3a65624b0a465eb6d47ce6237c678829f6).